crates/cheatcodes/src/inspector.rs

```diff
@@ -590,23 +590,33 @@ impl<DB: DatabaseExt + Send> Inspector<DB> for Cheatcodes {
     }
 
     fn step_end(&mut self, interpreter: &mut Interpreter<'_>, data: &mut EVMData<'_, DB>) {
-        if self.use_zk_vm && interpreter.current_opcode() == opcode::BALANCE {
-            if interpreter.stack.is_empty() {
-                interpreter.instruction_result = InstructionResult::StackUnderflow;
-                return;
-            }
-            // Safety: Length is checked above.
-            let address = Address::from_word(B256::from(unsafe { interpreter.stack.pop_unsafe() }));
-            let balance = foundry_zksync_core::balance(address, data.db, &mut data.journaled_state);
-
-            // Skip the current BALANCE instruction since we've already handled it
-            match interpreter.stack.push(balance) {
-                Ok(_) => unsafe {
-                    interpreter.instruction_pointer = interpreter.instruction_pointer.add(1);
-                },
-                Err(e) => {
-                    interpreter.instruction_result = e;
+        if !self.use_zk_vm {
+            return;
+        }
+
+        let address = match interpreter.current_opcode() {
+            opcode::SELFBALANCE => interpreter.contract().address,
+            opcode::BALANCE => {
+                if interpreter.stack.is_empty() {
+                    interpreter.instruction_result = InstructionResult::StackUnderflow;
+                    return;
                 }
+
+                Address::from_word(B256::from(unsafe { interpreter.stack.pop_unsafe() }))
+            }
+            _ => return,
+        };
+
+        // Safety: Length is checked above.
+        let balance = foundry_zksync_core::balance(address, data.db, &mut data.journaled_state);
+
+        // Skip the current BALANCE instruction since we've already handled it
+        match interpreter.stack.push(balance) {
+            Ok(_) => unsafe {
+                interpreter.instruction_pointer = interpreter.instruction_pointer.add(1);
+            },
+            Err(e) => {
+                interpreter.instruction_result = e;
```

Comment:
Perhaps we can make this a closure and have the logic represented as foundry does.

```rs
if self.use_zksync {
  let get_zksync_balance = |address: Address| {
      // Safety: Length is checked above.
        let balance = foundry_zksync_core::balance(address, data.db, &mut data.journaled_state);

        // Skip the current BALANCE instruction since we've already handled it
        match interpreter.stack.push(balance) {
            Ok(_) => unsafe {
                interpreter.instruction_pointer = interpreter.instruction_pointer.add(1);
            },
            Err(e) => {
                interpreter.instruction_result = e;
       }
  }

  match interpreter.current_opcode() {
            opcode::SELFBALANCE => {
                        get_zksync_balance(interpreter.contract().address)
            },
            opcode::BALANCE => {
                if interpreter.stack.is_empty() {
                    interpreter.instruction_result = InstructionResult::StackUnderflow;
                    return;
                }

                let address = Address::from_word(B256::from(unsafe { interpreter.stack.pop_unsafe() }))
                get_zksync_balance(address);
            }
            _ => return,
        };
}
```